### PR TITLE
deps/media-playback: Rewrite to use queue and states

### PR DESCRIPTION
**Pull request by nkmerrill**
Media source object is rewritten in to use states and a message queue.
 By moving to a state enumerator, the state can be guaranteed through
 each iteration of the message loop. This should greatly reduce race
 condition concerns. Also broke each message into its own function that
 can be logically modified to change behavior later.

 Additionally, a message queue is created to handle events from other
 threads. A function pointer allows new states to be added easily. The
 queue also features a bit flag to ensure any given state only appears
 once.

 Makes code more readable, will make future feature implementation
 easier, and provides a more efficient execution loop.

### Description
This PR includes two key changes to the media source file:
First, states are changed from being a bunch of flags to a discrete state enumerator. This makes it easy to understand what is happening at any given time, and makes code much more readable.
Second, events are now handled by a message queue. This queue is fairly standard, but has the unique feature of providing a flag table to ensure any given message is pushed only once. 

### Motivation and Context
Seeks to improve code readability and efficiency of media source loop.

### How Has This Been Tested?
Tested any combination of media events, and provided properties in the media source settings panel. Tested multiple sources and scenes. 

Tested with Windows 10 18363 on a AMD Ryzen 3900x.

Changes should be encapsulated to the media file, but the event signalers may be relevant to obs-ffmpeg-source if the new structure changes the logic in an unanticipated way.

### Types of changes

 - Performance enhancement (non-breaking change which improves efficiency)
 - Code cleanup (non-breaking change which makes code smaller or more readable)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: deps/media-playback/media-playback/media.h [hunk 1]
```diff
@@ -38,11 +38,66 @@ extern "C" {
...
+
+
+/*Used internally by mp_media to track current state.
+  to add new event:
+  1. Declare event in mp_media_state enumerator
+  2. Define event handler function.
+  3. Register event handler reference to queue.
+  4. Register any transition actions to the state change method.
+  Then use an invoke function to push the message to queue.
+  Note: States are mapped to an integer through a 2^x function,
```
Comment (kkartaltepe) [inline deps/media-playback/media-playback/media.h:57]: Is there a purpose behind using a magic function (that's not actually 2^x as evidenced by the code that does this mapping) instead of assigning the appropriate values directly in the enum?

Comment (nkmerrill) [reply deps/media-playback/media-playback/media.h:57]: I see your point. My thinking was that it would make implementing new states easier and avoid potential issues from typos, but this is adding a lot of computations for a management item that is not particularly difficult or complex.

File: deps/media-playback/media-playback/media.c [hunk 2]
```diff
@@ -481,55 +482,45 @@ static void seek_to(mp_media_t *m, int64_t pos)
...
 		mp_decode_flush(&m->a);
 }
 
-static bool mp_media_reset(mp_media_t *m)
+static bool mp_media_reset(mp_media_t *m, enum mp_media_state oldstate)
 {
-	bool stopping;
-	bool active;
-
+	/*Seeks to beginning of stream*/
```
Comment (RytoEX) [inline deps/media-playback/media-playback/media.c:495]:
```
// Use this style for single line comments.
/* Or use this style for single line comments. */
```

File: deps/media-playback/media-playback/media.c [hunk 3]
```diff
@@ -481,55 +482,45 @@ static void seek_to(mp_media_t *m, int64_t pos)
...
-	if (active) {
+	if (oldstate == mstate_stopped) {
+		m->start_ts = m->next_pts_ns = mp_media_get_next_min_pts(m);
+		m->play_sys_ts = (int64_t)os_gettime_ns();
+		m->next_ns = 0;
+
+		if (m->is_local_file && m->v_preload_cb)
+			mp_media_next_video(m, true);
+	} else {
+
```
Comment (RytoEX) [inline deps/media-playback/media-playback/media.c:517]: Remove this extra blank line.

File: deps/media-playback/media-playback/media.c [hunk 4]
```diff
@@ -653,93 +637,349 @@ static bool init_avformat(mp_media_t *m)
...
+		m->stop_cb(m->opaque);
+
+	return true;
+}
+
+/*Flip the looping flag*/
+static inline bool mp_media_setloop(mp_media_t *m)
+{
+	enum mp_media_state oldstate = m->mp_state;
+
```
Comment (RytoEX) [inline deps/media-playback/media-playback/media.c:740]: Remove the extra blank newlines in this function.  A single empty newline before the return statement should suffice.

File: deps/media-playback/media-playback/media.c [hunk 5]
```diff
@@ -653,93 +637,349 @@ static bool init_avformat(mp_media_t *m)
...
-		pthread_mutex_unlock(&m->mutex);
+/*Executes after leaving a state. Keep in mind that this won't execute
+  until AFTER the entry method of the new state. Therefore, if cleanup
+  is needed before changing states, put that logic in the entry
+  function of the new state.*/
+static inline void mp_media_change_state(mp_media_t *m,
+					 enum mp_media_state oldstate)
+{
+	switch (oldstate) {
+	case mstate_uninit:
```
Comment (RytoEX) [inline deps/media-playback/media-playback/media.c:805]: If restructuring this into a switch statement, remove the blank lines between the `case` and `break` statements.

File: deps/media-playback/media-playback/media.c [hunk 6]
```diff
@@ -653,93 +637,349 @@ static bool init_avformat(mp_media_t *m)
...
+			break;
+		default:
+			mp_media_push_message(m, mstate_playing);
+			break;
+		}
+	}
+	return true;
+}
+
+/*Executes when stop event occurs. One time action,
```
Comment (RytoEX) [inline deps/media-playback/media-playback/media.c:714]:
```
/* Use this style
 * for multi-line
 * comments.
 */
```
See [Section 8 of the Linux kernel coding style](https://github.com/torvalds/linux/blob/2324d50d051ec0f14a548e78554fb02513d6dcef/Documentation/process/coding-style.rst#8-commenting) which is linked from [our Contributing Guidelines](https://github.com/obsproject/obs-studio/blob/7031e2ec3a6ab2f97e7d8f50640d16b306f01e59/CONTRIBUTING.rst).

File: deps/media-playback/media-playback/media.c [hunk 7]
```diff
@@ -653,93 +637,349 @@ static bool init_avformat(mp_media_t *m)
...
+/*Executes the top message in the queue.*/
+static bool mp_media_pop_message(mp_media_t *m)
+{
+	bool result = true;
+
+	mp_media_queue_node_t *front = m->mp_queue.first;
+	/*Only execute if a state change actually occurs. If an
+	  action needs to be called within the state, add a new
+	  state that executes the action then moves back.*/
+	if (m->mp_state != front->message.state) {
```
Comment (RytoEX) [inline deps/media-playback/media-playback/media.c:850]: No need for curly braces for a single-line if statement.

File: deps/media-playback/media-playback/media.c [hunk 8]
```diff
@@ -653,93 +637,349 @@ static bool init_avformat(mp_media_t *m)
...
+		result = front->message.state_handler(m);
+	}
+
+	pthread_mutex_lock(&m->mutex);
+
+	if (m->mp_queue.first->next != NULL)
```
Comment (RytoEX) [inline deps/media-playback/media-playback/media.c:856]: Don't mix no-curly-braces blocks with curly-braces blocks in the same statement.

See [Section 3 of the Linux kernel coding style](https://github.com/torvalds/linux/blob/2324d50d051ec0f14a548e78554fb02513d6dcef/Documentation/process/coding-style.rst#3-placing-braces-and-spaces) which is linked from [our Contributing Guidelines](https://github.com/obsproject/obs-studio/blob/7031e2ec3a6ab2f97e7d8f50640d16b306f01e59/CONTRIBUTING.rst).

File: deps/media-playback/media-playback/media.c [hunk 9]
```diff
@@ -653,93 +637,349 @@ static bool init_avformat(mp_media_t *m)
...
+
+static bool mp_media_push_message(mp_media_t *m, enum mp_media_state state)
+{
+	/*To prevent the queue from growing infinitely, and to
+	  prevent needless repetition, use a bit flag.
+	  Queue should only contain any given message once.*/
+	if (m->mp_queue.flags & state)
+		return false;
+
+	mp_media_queue_node_t *newnode = malloc(sizeof(mp_media_queue_node_t));
```
Comment (RytoEX) [inline deps/media-playback/media-playback/media.c:878]: I'd suggest using either `new_node` or `node`.  Not sure the point of using "new" since there is no other variable in this scope named `node`.

File: deps/media-playback/media-playback/media.c [hunk 10]
```diff
@@ -653,93 +637,349 @@ static bool init_avformat(mp_media_t *m)
...
+	mp_media_state_message_t *newmessage;
```
Comment (RytoEX) [inline deps/media-playback/media-playback/media.c:879]: I'd suggest using either `new_message` or `message`.  Not sure the point of using "new" since there is no other variable in this scope named `message`.

File: deps/media-playback/media-playback/media.c [hunk 11]
```diff
@@ -653,93 +637,349 @@ static bool init_avformat(mp_media_t *m)
...
+
+/*Manage loop timing.*/
+static inline bool mp_media_tick(mp_media_t *m)
+{
+	/*If there is an event that should block the loop, place it
+	  here. Make sure to address semaphore logic in all events.
+	  For instance, a stop invoke should check for a pause event
+	  and signal only then. If stop always signals, then that
+	  signal may prevent the block for stop itself, but without
+	  signalling out of a pause, the stop event will not run until
```
Comment (RytoEX) [inline deps/media-playback/media-playback/media.c:945]: Not a big deal, but "signaling" with one "l" is the more commonly accepted American English spelling.

File: deps/media-playback/media-playback/media.h [hunk 12]
```diff
@@ -38,11 +38,65 @@ extern "C" {
 #pragma warning(pop)
 #endif
 
+
```
Comment (RytoEX) [inline deps/media-playback/media-playback/media.h:41]: Don't add extra blank newlines unless there's a true need.

**Comment by nkmerrill** [comment]
Rebased and applied formatting changes

**Comment by nkmerrill** [comment]
Rebased, reviewed and adjusted formatting again.

File: deps/media-playback/media-playback/media.c [hunk 13]
```diff
@@ -749,17 +960,22 @@ static void *mp_media_thread_start(void *opaque)
...
+	if (!mp_media_thread(m))
+		if (m->stop_cb) 
 			m->stop_cb(m->opaque);
-		}
-	}
 
 	return NULL;
 }
 
-static inline bool mp_media_init_internal(mp_media_t *m,
```
Comment (RytoEX) [inline deps/media-playback/media-playback/media.c:761]: Why did you remove the `inline` from `mp_media_init_internal`?

Comment (nkmerrill) [reply deps/media-playback/media-playback/media.c:761]: I couldn't see why I did either, so I reverted the change. Tested and don't see any issues.

**Comment by nkmerrill** [comment]
Additional format changes, made comments more consistent, and fixed commit title.

File: deps/media-playback/media-playback/media.c [hunk 14]
```diff
@@ -754,12 +966,18 @@ static void *mp_media_thread_start(void *opaque)
 			m->stop_cb(m->opaque);
 		}
 	}
-
```
Comment (RytoEX) [inline deps/media-playback/media-playback/media.c:757]: Add this blank line back. You don't need to remove pre-existing blank lines like this. Extra whitespace changes just creates extra noise in the diff.

Comment (nkmerrill) [reply deps/media-playback/media-playback/media.c]: Done.

File: deps/media-playback/media-playback/media.h [hunk 15]
```diff
@@ -41,8 +41,62 @@ extern "C" {
 typedef void (*mp_video_cb)(void *opaque, struct obs_source_frame *frame);
 typedef void (*mp_audio_cb)(void *opaque, struct obs_source_audio *audio);
 typedef void (*mp_stop_cb)(void *opaque);
+typedef bool (*mp_state_handler)(struct mp_media_t *m);
```
Comment (jp9000) [inline deps/media-playback/media-playback/media.h:44]: This causes various warnings in VC, because `struct mp_media_t` is an unknown type.  Plus, mp_media_t is not a struct (technically it is but I'll get to that), it's a typedef. So you should not be specifying `struct mp_media_t`.

To fix this, just do a forward declare somewhere above this:
```C
struct mp_media;
typedef struct mp_media mp_media_t;
```
to be in accordance with the structure itself below, then change this line to:
```C
typedef bool (*mp_state_handler)(mp_media_t *m);
```

File: deps/media-playback/media-playback/media.c [hunk 16]
```diff
@@ -779,7 +998,13 @@ static inline bool mp_media_init_internal(mp_media_t *m,
 		return false;
 	}
 
+	pthread_mutex_lock(&m->mutex);
+
 	m->thread_valid = true;
+	mp_media_init_queue(&m);
```
Comment (jp9000) [inline deps/media-playback/media-playback/media.c:1004]: You are initializing the pointer to the pointer, which causes invalid data to changed.  This is specified in the warning. 
This should be:
```C
mp_media_init_queue(m);
```

**Review by jp9000 — CHANGES_REQUESTED** [review]
Because of the push/pop mechanism, this causes seeking to occasionally not unpause playback when you stop seeking.  The "unpause" operation will get queued, then the seek operation may complete after that, causing it to queue a "pause" operation, which will get queued *after* the "unpause" operation, causing it to stay paused.  This is basically a race condition without having any actual data races, and is mostly a flaw in the design of the circular queue system itself.  A solution to fix this is not by checking the queue to make sure there are no pause operations, that would make me feel very uncomfortable.  The solution to fix this is likely to separate certain states from one another, to limit what states are queued, and/or to limit where they're queued from.  That's sort of why I had separate variables in the first place.

It makes me feel very uncomfortable that a circular queue running in a thread of its own has so many possible states, and that the thread itself can push states at any time which may conflict with yet-to-be-processed states that were already pushed by another thread.

I think you need to really take another look at this design, and try to gauge more of its possible follies like that.  The one I pointed out is the only one I've seen so far, but maybe you should really examine whether all these states are actually necessary or need to be assigned to one single master variable that's changed by a circular queue.

I'm probably making it sound like more of a big deal than it actually is, but when it comes to threading, I am very paranoid.

Also, if you don't have access to a visual studio compiler make sure to check the visual studio build output for warnings.  The visual studio compiler despite its faults has quite a number of different warnings that gcc/clang might not spot, and I'm guessing that's why you didn't see them.  Although you really shouldn't consider the work complete when there's warnings in the first place.

**Comment by nkmerrill** [comment]
Converting to draft for 2 reasons:
1. We have already fixed the mentioned bugs with the current media source implementation so this is no longer a bug fix. I will continue to work on this as an enhancement project seeking to improve efficiency and readability of the media source loop.
2. I need to address the concerns regarding potential race conditions created by the queue system being used internally by the loop for cleanup. This can create race conditions as the loop may have a messaged pushed to it in an undesired order. State change cleanups need to be handled internally before the queue is popped again.

**Comment by WizardCM** [comment]
Hi @nkmerrill - I was wondering if you have any updates on the status of this PR.

**Comment by nkmerrill** [comment]
> 
> 
> Hi @nkmerrill - I was wondering if you have any updates on the status of this PR.

Hi @WizardCM, I haven't had as much time to work on this I'd like, but I am still looking into this. Thanks for checking in.

**Comment by Lain-B** [comment]
I had some nitpicks about this one which is why I didn't end up merging this, but for now I'm going to close this, and I'm going to have to implement queues myself at some point. The reason I'm going to just have to do it myself at some point is because I had to significantly rework some of the media playback code without merging this, and because of that it basically invalidates all of this code, and  I can't ask anyone to redo their work around that.

We do need queues for it, but it'll just have to wait a bit.
